a few changes towards tensorflow 2 compatibility

Bomme opened the pull request:
This PR removes `tf.contrib` from the codebase to make it possible to use tensorflow 2.

Changes to note:
1) introduce [tf-slim](https://github.com/google-research/tf-slim) for `vggish`. This model definition is now more in line with the original implementation in https://github.com/tensorflow/models/blob/master/research/audioset/vggish/vggish_slim.py
2) Switch from [contrib.layers.variance_scaling_initializer](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/contrib/layers/variance_scaling_initializer) to [tf.compat.v1.keras.initializers.VarianceScaling](https://www.tensorflow.org/api_docs/python/tf/compat/v1/keras/initializers/VarianceScaling). I tried to match the initializers behaviour and adapted the to the best of my knowledge.
EDIT: We finally settled for `tensorflow.compat.v1.initializers.variance_scaling` since `keras.initializers.VarianceScaling` gave different random numbers.
3) Rely on `import tensorflow.compat.v1 as tf` and `tf.disable_v2_behavior()` for now. If we want to fully migrate we can move individual functions to the new API.
4) Greatly simplify `requirements.txt` and allow tensorflow 2

Bomme committed:
Use tensorflow.compat.v1 more consistently

Bomme committed:
switch to tf.keras.initializers.VarianceScaling

this initializer might not behave exactly as the old but I tried to match all the parameters

Bomme committed:
switch to initializers.VarianceScaling in classification_heads.py and use compat.v1

this initializer might not behave exactly as the old but I tried to match all the parameters

Bomme committed:
switch vggish model to tf-slim

The vggish model is now more aligned with the original code in the
tensorflow models repository

Bomme commented on src/models_transfer_learning.py:
```diff
@@ -1,5 +1,9 @@
 import tensorflow as tf
```
@palonso Should we move the whole file also to the `tensorflow.compat.v1`?

palonso replied:
Yes, otherwise we will have problems with `disable_v2_behavior()`

Bomme replied:
updated the call to `disable_v2_behavior()`. We can simplify later if we need to.

palonso replied:
there are more TF1 objects that are only available under the `tf.compat.v1.` scope (e.g., `truncated_normal_initializer`, `variable_scope`...).

The quick solution would be,
`import tensorflow as tf` → `import tf.compat.v1 as tf`

Otherwise, we have to identify all the objects that require `tf.compat.v1` and modify them as `disable_v2_behavior()`

Bomme committed:
fix call to compat.v1.disable_v2_behavior() in models_transfer_learning

palonso commented on src/classification_heads.py:
```diff
@@ -1,31 +1,40 @@
-                                    activation=None,
-                                    units=config['num_classes_dataset'],
-                                    kernel_initializer=tf.contrib.layers.variance_scaling_initializer(seed=config['seed']))
+        initializer = tf.keras.initializers.VarianceScaling(scale=2.0, seed=config["seed"])
```
To keep the behaviour of the [VarianceScaling](https://www.tensorflow.org/api_docs/python/tf/keras/initializers/VarianceScaling) initializer we should set the parameter `distribution` to `untruncated_normal`, which was the default behaviour of [tf.contrib.layers.variance_scaling_initializer](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/contrib/layers/variance_scaling_initializer). Although on informal testing `truncated_normal` seems to be helping on training and it is interesting to explore its impact.

Bomme replied:
I might be missing something, but I understood the [documentation](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/contrib/layers/variance_scaling_initializer) exactly in the opposite way:

> When initializing a deep network, it is in principle advantageous to keep the scale of the input variance constant, so it does not explode or diminish by reaching the final layer. This initializer use the following formula:
```  if mode='FAN_IN': # Count only number of input connections.
    n = fan_in
  elif mode='FAN_OUT': # Count only number of output connections.
    n = fan_out
  elif mode='FAN_AVG': # Average number of inputs and output connections.
    n = (fan_in + fan_out)/2.0

    truncated_normal(shape, 0.0, stddev=sqrt(factor / n))
```

This made me think that the default was `truncated_normal`. 
And in the code I found the same thing:
https://github.com/tensorflow/tensorflow/blob/v1.15.0/tensorflow/contrib/layers/python/layers/initializers.py#L149

palonso replied:
Nice! After looking at the code I think you are right, the value should be `truncated_normal` (the default). 
Even though I was facing diverging results, so I did an additional experiment:

IN TF 1.15:
```Python
ini = tf.contrib.layers.variance_scaling_initializer(seed=6)
with tf.compat.v1.Session() as sess:  print(ini([10]).eval())
```
`[ 0.3352677, 0.3420339, 0.18761884, 0.1507059, 0.56236863, -0.46897727, 0.15543926, -0.33059058, -0.71044177, 0.14339857]`

In TF 2.3.1

```Python
ini = tf.keras.initializers.VarianceScaling(scale=2.0, seed=6)
ini([10])
```
`[ 0.6524273, -0.2395821,  0.2508376, -0.11647213, -0.6710456 , 0.74666166,  0.9594222,  0.16396938,  0.4017174,  0.3574336 ]`

An then,
```Python
ini = tf.compat.v1.initializers.variance_scaling(scale=2.0, seed=6)
ini([10])
```
`[ 0.33428904,  0.3410355 ,  0.18707119,  0.15026599,  0.56072706, -0.46760836,  0.15498553, -0.3296256 , -0.708368,  0.14298]`

The last method returns almost the same values as in TF 1.15, so I think we should use it instead.
I've also check it in my experimental setup and the results are closer to the ones reported with TF 1.15

However, I couldn't find a satisfactory explanation. The implementations seem very similar ([v1](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/init_ops.py), [v2](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/init_ops_v2.py)).

Maybe the only difference is the random engine?

Bomme replied:
Good work! I must have missed `tf.compat.v1.initializers.variance_scaling` as a possiblitliy! 
I adapted to use the `compat` function!

palonso reviewed (changes requested):
https://github.com/palonso/musicnn-training/blob/ce1013529bfb3797677490831c6d5467dd1e6936/requirements.txt#L31

Replace `tensorflow==1.15.0` →  `tensorflow>=1.15.0` 
so that a previous installation of TF 2.X is not replaced on the `pip install -r requirements.txt` step.

Bomme committed:
make tf.compat.v1 the default in models_transfer_learning.py

Bomme committed:
for reproducibility stick to tf.compat.v1.initializers.variance_scaling

in small tests we saw that keras.initializers.VarianceScaling gave
different random numbers than the original code even when specifying the
random seed. This compat layer gives almost the same result
(differences < 0.01)

Bomme committed:
reduce requirements.txt to only direct dependencies and comment others

This is my best effort to find the true dependencies of this project.
With a minimal requirements.txt it might be easier to upgrade single
dependencies

Bomme committed:
simplify requirements.txt: delete commented lines

Bomme committed:
sort requirements.txt alphabetically

Bomme committed:
allow tensorflow 2 in requirements.txt

Bomme committed:
relaxing requirements on dependencies by allowing all bugfix versions

Bomme commented:
> https://github.com/palonso/musicnn-training/blob/ce1013529bfb3797677490831c6d5467dd1e6936/requirements.txt#L31
> 
> Replace `tensorflow==1.15.0` → `tensorflow>=1.15.0`
> so that a previous installation of TF 2.X is not replaced on the `pip install -r requirements.txt` step.

I greatly simplified `requirements.txt` and relaxed the versions a bit. That should be enough for tensorflow 2.0. For later versions we might need to look into further relaxing the requirements for `numpy` & `scipy`.

Bomme committed:
inline initializers when they are called only once

Bomme committed:
bring tf.compat.v1 & variance_scaling to more of the training scripts

Bomme committed:
replace contrib.layers.flatten with compat.v1.layers.flatten
